File: go/ch03/toy_score.go
// go/ch03/toy_score.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type ToyState struct {
	Picked []bool
}

type ToyInput struct {
	Values []int
}

func applyMove(state ToyState, index int, pick bool) ToyState {
	nxt := append([]bool{}, state.Picked...)
	nxt[index] = pick
	return ToyState{nxt}
}

func score(state ToyState, inp ToyInput) int {
	total := 0
	for i, picked := range state.Picked {
		if picked {
			total += inp.Values[i]
		}
	}
	return total
}

func main() {
	sc := bufio.NewScanner(os.Stdin)
	sc.Scan()
	n, _ := strconv.Atoi(sc.Text())
	sc.Scan()
	vals := parseInts(sc.Text())
	inp := ToyInput{vals}
	state := ToyState{Picked: make([]bool, n)}

	for i := 0; i < n; i++ {
		state = applyMove(state, i, true)
		fmt.Printf("day %d: score=%d\n", i+1, score(state, inp))
	}
}

func parseInts(line string) []int {
	parts := strings.Fields(line)
	out := make([]int, len(parts))
	for i, p := range parts {
		out[i], _ = strconv.Atoi(p)
	}
	return out
}

File: python/ch03/toy_score.py
# python/ch03/toy_score.py
"""最小例: State / Move / Score のモデリング."""
from dataclasses import dataclass
from typing import List


@dataclass
class ToyState:
    picked: List[bool]


@dataclass
class ToyInput:
    values: List[int]


def apply_move(state: ToyState, index: int, pick: bool) -> ToyState:
    nxt = ToyState(list(state.picked))
    nxt.picked[index] = pick
    return nxt


def score(state: ToyState, inp: ToyInput) -> int:
    total = 0
    for i, picked in enumerate(state.picked):
        if picked:
            total += inp.values[i]
    return total


def main() -> None:
    import sys

    lines = sys.stdin.read().splitlines()
    n = int(lines[0])
    values = list(map(int, lines[1].split()))
    inp = ToyInput(values)
    state = ToyState([False] * n)

    # デモ: 全部 pick する Move を順に適用
    for i in range(n):
        state = apply_move(state, i, True)
        print(f"day {i + 1}: score={score(state, inp)}")


if __name__ == "__main__":
    main()
